### PR TITLE
ilasm support for methods with multiple documents

### LINKED ISSUE #102198: ilasm produces wrong SequencePoints when a method references several Documents
### Description

When a method contains `.line` directives referring to several documents, `ilasm` ignores all but the last specified document.
This in turn impacts tools (such as Visual Studio) that will, for example, show the wrong file (and therefore an irrelevant line) when stepping through the built code.

### Reproduction Steps

After having built the tests, run the following program.

```
using System.Reflection;
using System.Reflection.Metadata;
using System.Reflection.PortableExecutable;

//Adapt the following two lines to suit your platform/configuration/folders.
var dotnetRuntimeRepo = "C:\\repositories\\dotnet\\runtime";
var pepath = $"{dotnetRuntimeRepo}\\artifacts\\tests\\coreclr\\windows.x64.Debug\\ilasm\\PortablePdb\\IlasmPortablePdbTests\\TestFiles\\TestDocuments1_win.dll";
//compiled from <dotnetRuntimeRepo>\src\tests\ilasm\PortablePdb\TestFiles\TestDocuments1_win.il
var pereader = new PEReader(File.OpenRead(pepath));
pereader.TryOpenAssociatedPortablePdb(pepath, s => File.OpenRead(s), out var pdbreader, out var pdbpath);
if (pdbreader == null) return;
var reader = pdbreader.GetMetadataReader();
foreach (var mdi in reader.MethodDebugInformation)
{
    if (mdi.IsNil) continue;
    foreach (var sp in reader.GetMethodDebugInformation(mdi).GetSequencePoints())
    {
        var d = sp.Document;
        if (d.IsNil) continue;
        Console.WriteLine(reader.GetString(reader.GetDocument(d).Name));
    }
}
```

For reference, here is TestsDocuments1_win.il:

```
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

.assembly extern System.Runtime
{
}

.assembly TestDocuments
{
}

.class public auto ansi beforefieldinit TestDocumentsType
       extends [System.Runtime]System.Object
{
    .method public hidebysig instance void Foo() cil managed
    {
    .line 1,1 : 9,10 'C:\\tmp\\non_existent_source1.cs'
        nop
    .line 2,2 : 9,10 'C:\\tmp\\non_existent_source2.cs'
        ret
    }
}

```

### Expected behavior

The following output is expected (on Windows, on other platforms the paths will be different but the important digit would be the same):

```
C:\tmp\non_existent_source1.cs
C:\tmp\non_existent_source2.cs
```

### Actual behavior

The following output is produced (again, on Windows with slight changes to the paths but not the digits on other platforms):

```
C:\tmp\non_existent_source2.cs
C:\tmp\non_existent_source2.cs
```

### Regression?

This is a functional regression compared to .NET Framework although it used a different (non-portable) PDB file format.

### Known Workarounds

Using .NET Framework where practical.

### Configuration

Observed on:
 
.NET 6 -> .NET 9 (preview 3)
Windows 10 x64, Windows 11 x64

Not specific to that configuration

### Other information

The issue is in the method https://github.com/dotnet/runtime/blob/039d2ecb46687e89337d6d629c295687cfe226be/src/coreclr/ilasm/portable_pdb.cpp#L201 which doesn't deal with an initial document (L208) and only provides a unique document (L314, notice that it uses `currSeqPoint` outside the loop).

I have a fix and a test for this issue that will be in an upcoming PR.

### PR DESCRIPTION
**Pull request by yjaradin**
Make sure the SequencePoints produced in the PDB
reference the documents indicated on the .line directive.

Fix #102198

### CODE REVIEW

**Comment by yjaradin** [comment]
@dotnet-policy-service agree company="Raincode"

**Review by ivanpovazan — COMMENTED** [review]
Thank you very much for reporting and fixing this (and adding the test case 😃) .
This already looks pretty good, I only left a few comments.

I also manually rerun the CI which is green now.

@jkotas @BruceForstall you might want to take a look.

File: src/coreclr/ilasm/portable_pdb.cpp
Comment (yjaradin) [reply src/coreclr/ilasm/portable_pdb.cpp:329]: Because that's what the spec says :)
https://github.com/dotnet/runtime/blob/main/docs/design/specs/PortablePdb-Metadata.md#methoddebuginformation-table-0x31
I think the idea is to not confuse tools that cannot handle multiple documents / not change the semantic from only document to first document.

**Comment by jkotas** [comment]
Original PR with portable PDB support: https://github.com/dotnet/runtime/pull/37702

**Review by noahfalk — COMMENTED** [review]
This looks pretty good to me, one question inline.

**Comment by noahfalk** [comment]
Thanks for a clear issue and well executed PR @yjaradin!

File: src/tests/ilasm/PortablePdb/IlasmPortablePdbTesterCommon.cs
Comment (ivanpovazan) [reply src/tests/ilasm/PortablePdb/IlasmPortablePdbTesterCommon.cs]: I understand and thanks for keeping it clean! 
I simplified the code a bit and added this small change to this PR (hope you are ok with that).

File: src/coreclr/ilasm/portable_pdb.cpp
Comment (yjaradin) [reply src/coreclr/ilasm/portable_pdb.cpp:229]: Not in any realistic way.

This line is conditioned on method->m_LinePCList.COUNT() != 0 and the only way that can be true is the PUSH in https://github.com/dotnet/runtime/pull/102199/files/0a2f49bb7de5d1432b4d1bbcc5174ff26e0d910d#diff-1f5d27b028692df005d7f33a02b3080f52ffb3a4e1384b906a7bad7186f5e601R1444 which can't be reached without setting m_FirstDocument to instr->pOwnerDocument. This one in turn is set from the m_currentDocument set by PortablePdbWriter::DefineDocument called by Assembler::SetSourceFileName itself called directly from wmain with the starting file name. The m_currentDocument can only be reset to null by calling the PortablePdbWriter::Init method which is never called after constructing the object.

SetSourceFileName does nothing if the passed string is null or start with a zero character (empty string) but this cannot be the case for the invocation from wmain as the file name is forced by MakeProperSourceFileName to contain at least a dot (it would add ".il" to it if it didn't).

This leaves the very convoluted case : on a machine with signed characters, separate representation for +0 and -0 (i.e. not a two's complement machine) where the C++ ABI define only one of them as a string terminator (I didn't dig in the C++ standard to see if it would even be conformant) and where the other can be the first character of a file name, SetSourceFileName would ultimately leave a null value pass. I doubt this would be the only place that fails in such a corner case.

File: src/tests/ilasm/PortablePdb/IlasmPortablePdbTesterCommon.cs
Comment (yjaradin) [reply src/tests/ilasm/PortablePdb/IlasmPortablePdbTesterCommon.cs]: Thanks a lot!

**Review by ivanpovazan — APPROVED** [review]
LGTM, thank you!

Although I would suggest someone else from the owners to approve before merging this.

**Review by noahfalk — APPROVED** [review]

**Comment by noahfalk** [comment]
Are you ready for us to merge this?

**Comment by ivanpovazan** [comment]
@noahfalk are we just waiting on @yjaradin confirmation?

**Comment by noahfalk** [comment]
> @noahfalk are we just waiting on @yjaradin confirmation?

Yeah. 
@yjaradin - just wanted to confirm this is ready for merge?

**Comment by yjaradin** [comment]
Sorry, I had been a bit side-tracked. Yes, this is ready for merge from my point of view.

**Pull request merged**
